pkg/metrics/instances.go

```diff
@@ -0,0 +1,105 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/otel"
+
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/global"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"go.opentelemetry.io/otel/attribute"
+
+	v1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
+)
+
+var groups = make(map[string]*instancesCounter)
+
+type instancesCounter struct {
+	Labels []attribute.KeyValue
+	Count  int64
+}
+type instancesMetric struct {
+	client client.Client
+	groups map[string]*instancesCounter
+}
+
+func newInstancesMetric(client client.Client) *instancesMetric {
+	return &instancesMetric{
+		client: client,
+		groups: make(map[string]*instancesCounter),
+	}
+}
+
+func (i *instancesMetric) Setup(ctx context.Context) error {
+	tracer := otel.GetTracerProvider().Tracer(v1.CustomMetricsTracer)
+	ctx, span := tracer.Start(ctx, "setup-jaeger-instances")
+	defer span.End()
+	meter := global.Meter(meterName)
+	_, err := meter.NewInt64ValueObserver("operator_jaeger_instances", i.callback,
+		metric.WithDescription("Number of jaeger instances in cluster"),
+	)
+	return err
+}
+func (i *instancesMetric) agentMode(jaeger v1.Jaeger) string {
+	agent := string(jaeger.Spec.Agent.Strategy)
+	if agent == "" {
+		return "sidecar"
+	}
+	return agent
+}
+func (i *instancesMetric) storage(jaeger v1.Jaeger) string {
+	storage := string(jaeger.Spec.Storage.Type)
+	if storage == "" {
+		storage = "memory"
+	}
+	return strings.ToLower(storage)
+
+}
+func (i *instancesMetric) strategy(jaeger v1.Jaeger) string {
+	strategy := string(jaeger.Spec.Strategy)
+	if strategy == "" {
+		return "allinone"
```

Comment:
if this is the convention if strategy is empty then we should define a method on the `Spec` object to return a strategy and set the allinone as default.

Reply:
The same for memory

Reply:
and agent sidecar

Reply:
are we using `Defaulters` in this version of the operator? I started to do it on v2 but IIRC we don't use it here.

Reply:
True. We have a normalizer in this operator.